Implement Client Credentials Authentication

pkostrzewa opened the pull request:
- [x] The `Filter` should process requests for the (default) path `/oauth2/token` and check if credentials are available in the request.
- [X]  The `OAuth2ClientAuthenticationToken` should be passed to the `AuthenticationManager`.
- [x]  The `AuthenticationManager` should be composed of `OAuth2ClientAuthenticationProvider`.
- [x]  The `OAuth2ClientAuthenticationProvider` should use the `RegisteredClientRepository`.
- [x]  The `RegisteredClient` should be returned in a new `OAuth2ClientAuthenticationToken` if the authentication succeeds.
- [X] The `Filter` should save the  `OAuth2ClientAuthenticationToken` in the `SecurityContext`.
- [x] Add javadoc for `class` and `public` methods.
- [x] Unit tests.

The `ClientAuthenticationConverter` is based on `BasicAuthenticationFilter`.

Fixes #39 .

pkostrzewa commented:
@jgrandja Should `AuthenticationSuccessHandler` and `AuthenticationFailureHandler` be used in early implementation?

watsta commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientAuthenticationProvider.java:
```diff
@@ -15,20 +15,64 @@
+		RegisteredClient registeredClient = registeredClientRepository.findByClientId(clientId);
+		// https://tools.ietf.org/html/rfc6749#section-2.4
+		if (registeredClient == null) {
+			logger.debug("Authentication failed: client not registered");
```
wouldn't it be helpful to include the clientId in the debug message?

watsta commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientAuthenticationProvider.java:
```diff
@@ -15,20 +15,64 @@
+
+		String presentedSecret = authentication.getCredentials()
+				.toString();
+		if (!registeredClient.getClientSecret()
```
I think clients are not required to have a secret. Obviously, such clients should not be used with the client credentials flow, but that could be something to mitigate here. I think a simple null-check would make sense

watsta commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/web/ClientAuthenticationConverter.java:
```diff
@@ -0,0 +1,70 @@
+
+		String token = new String(decoded, StandardCharsets.UTF_8);
+		String[] credentials = token.split(":");
+		if (credentials.length == 0) {
```
shouldn't this rather be `credentials.length != 2`? Otherwise you can still get an `ArrayIndexOutOfBoundsException` on line 68, e.g when there is no `:` present in the Authorization header, length will be 1

pkostrzewa replied:
That's true.

nifrez commented:
@watsta It's draft at the moment ;)

Captain-P-Goldfish commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2ClientAuthenticationFilter.java:
```diff
@@ -16,25 +16,103 @@
+				.setAuthentication(result);
+	}
+
+	protected void onUnsuccessfulAuthentication(HttpServletRequest request, HttpServletResponse response,
```
renaming proposal: "onFailedAuthentication"
the current state made me misread this 2 times.

pkostrzewa replied:
Agree.

jgrandja replied:
Replace with `AuthenticationFailureHandler` delegate and add `final` setter to allow for customization.

pkostrzewa commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientAuthenticationProvider.java:
```diff
@@ -15,20 +15,64 @@
 public class OAuth2ClientAuthenticationProvider implements AuthenticationProvider {
 	private RegisteredClientRepository registeredClientRepository;
 
+	protected final Log logger = LogFactory.getLog(getClass());
```
I'd rather stick to that is already configured and present rather than adding slf4j in this PR.

pkostrzewa commented:
@jgrandja Could you please elaborate on "The `AuthenticationManager` should be composed of `OAuth2ClientAuthenticationProvider`?

Isn't `OAuth2ClientAuthenticationProvider` automatically registered in `AuthenticationManager`? If not, what should be done to register a provider?

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2ClientAuthenticationToken.java:
```diff
@@ -23,6 +23,7 @@
  * @author Joe Grandja
+ * @author Patryk Kostrzewa
  */
 public class OAuth2ClientAuthenticationToken extends AbstractAuthenticationToken {
```
For constructor arguments, apply `Assert.notNull()` or `Assert.hasText()` or whatever makes sense.

For private members, mark `final` whenever applicable (set in constructor and no corresponding setter)

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2ClientAuthenticationFilter.java:
```diff
@@ -16,25 +16,103 @@
 import java.io.IOException;
 
 /**
+ * Processes a HTTP request's BASIC authorization headers, putting the result into the
```
Please remove this javadoc that came from `BasicAuthenticationFilter` as not all details apply to this filter. For example, this `Filter` will support multiple authentication schemes (not just basic) via a composite `AuthenticationConverter`.

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2ClientAuthenticationFilter.java:
```diff
@@ -16,25 +16,103 @@
 public class OAuth2ClientAuthenticationFilter extends OncePerRequestFilter {
-	private AuthenticationManager authenticationManager;
+
+	private static final String TOKEN_ENDPOINT = "/oauth2/token";
```
The default token endpoint is "/oauth2/token" but it should also be configurable via a constructor. Take a look at `OAuth2AuthorizationRequestRedirectFilter` to see how a custom path can be supplied.

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2ClientAuthenticationFilter.java:
```diff
@@ -16,25 +16,103 @@
+
+	private static final String TOKEN_ENDPOINT = "/oauth2/token";
+	private final AuthenticationManager authenticationManager;
+	private final ClientAuthenticationConverter authenticationConverter = new ClientAuthenticationConverter();
```
Let's reference the interface `private AuthenticationConverter authenticationConverter` and provide a setter to enable customization `public final void setAuthenticationConverter(AuthenticationConverter authenticationConverter)`

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2ClientAuthenticationFilter.java:
```diff
@@ -16,25 +16,103 @@
 	@Override
-	protected void doFilterInternal(HttpServletRequest request,
-			HttpServletResponse response, FilterChain filterChain)
+	public void afterPropertiesSet() {
```
Please remove as `AuthenticationManager` is required as per constructor.

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2ClientAuthenticationFilter.java:
```diff
@@ -16,25 +16,103 @@
+	protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain)
 			throws ServletException, IOException {
 
+		if (TOKEN_ENDPOINT.equals(request.getRequestURI())) {
```
Let's use `RequestMatcher` specifically `AntPathRequestMatcher`. See `OAuth2AuthorizationRequestRedirectFilter` -> `DefaultOAuth2AuthorizationRequestResolver` -> `RequestMatcher`

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2ClientAuthenticationFilter.java:
```diff
@@ -16,25 +16,103 @@
+				onSuccessfulAuthentication(request, response, result);
+			} catch (AuthenticationException failed) {
+				onUnsuccessfulAuthentication(request, response, failed);
+				filterChain.doFilter(request, response);
```
Failed authentication should not proceed through the chain. The failure handler should respond with error response.

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2ClientAuthenticationFilter.java:
```diff
@@ -16,25 +16,103 @@
+		filterChain.doFilter(request, response);
 	}
 
+	protected void onSuccessfulAuthentication(HttpServletRequest request, HttpServletResponse response,
```
Replace with `AuthenticationSuccessHandler` delegate and add `final` setter to allow for customization.

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2ClientAuthenticationFilter.java:
```diff
@@ -16,25 +16,103 @@
+	protected void onUnsuccessfulAuthentication(HttpServletRequest request, HttpServletResponse response,
+			AuthenticationException failed) throws IOException {
+
+		SecurityContextHolder.clearContext();
```
We also need to write the OAuth 2.0 Error Response based on the incoming `OAuth2AuthenticationException`

pkostrzewa replied:
I've changed this to simply return 401 status for now in default `AuthenticationFailureHandler`. Should I write an `OAuth2Error` to `response` instead?

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/web/ClientAuthenticationConverter.java:
```diff
@@ -0,0 +1,70 @@
+ *
+ * @author Patryk Kostrzewa
+ */
+public class ClientAuthenticationConverter implements AuthenticationConverter {
```
Rename to `OAuth2ClientAuthenticationConverter`

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/web/ClientAuthenticationConverter.java:
```diff
@@ -0,0 +1,70 @@
+		}
+
+		if (header.equalsIgnoreCase(AUTHENTICATION_SCHEME_BASIC)) {
+			throw new BadCredentialsException("Empty basic authentication token");
```
We should throw `OAuth2AuthenticationException` using `invalid_client` as the `OAuth2Error.errorCode`. See [5.2.  Error Response](https://tools.ietf.org/html/rfc6749#section-5.2)

pkostrzewa replied:
Shouldn't the error be `invalid_request` in this case?

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/web/ClientAuthenticationConverter.java:
```diff
@@ -0,0 +1,70 @@
+			decoded = Base64.getDecoder()
+					.decode(base64Token);
+		} catch (IllegalArgumentException e) {
+			throw new BadCredentialsException("Failed to decode basic authentication token");
```
throw `OAuth2AuthenticationException `

jgrandja commented:
@pkostrzewa 

> Isn't `OAuth2ClientAuthenticationProvider` automatically registered in `AuthenticationManager`?

Yes, it will be in a later story. Nothing to do as far as this issue goes. I was simply stating the runtime expectations. I updated the main issue to be clear.

pkostrzewa commented:
@jgrandja 

Thanks for the review! I will update the PR with your comments later on today.

pkostrzewa commented:
@jgrandja I've updated the PR. There are still missing tests and documentation and I will provide both as soon as I can.

jgrandja commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/DefaultOAuth2ClientAuthenticationFailureHandler.java:
```diff
@@ -0,0 +1,39 @@
+			final AuthenticationException exception) throws IOException, ServletException {
+
+		SecurityContextHolder.clearContext();
+		response.sendError(HttpStatus.UNAUTHORIZED.value());
```
We should also write out the `OAuth2Error` in the response. Try to reuse `OAuth2ErrorHttpMessageConverter` from `spring-security-oauth2-client` module.

jgrandja commented on core/src/test/java/org/springframework/security/oauth2/server/authorization/web/OAuth2ClientAuthenticationConverterTests.java:
```diff
@@ -0,0 +1,116 @@
+	}
+
+	@Test
+	public void shouldConvertToAuthenticationToken() {
```
Please see existing tests in `spring-security-oauth2-client` module and follow the method naming pattern and general test style. eg. `DefaultOAuth2AuthorizationRequestResolverTests`, `OAuth2AuthorizationCodeAuthenticationProviderTests`, etc.

pkostrzewa commented:
Thanks for the review @jgrandja. I will update the PR soon.

pkostrzewa commented:
> For the next update, can you please rebase on `master` and avoid merge commits. And after you add all the tests please change the Draft to Ready and I think we'll be close to merging.

Do you mean I should squash all commits from this PR into single one?

jgrandja commented:
@pkostrzewa Yes please squash to 1 commit and rebase on latest master.

pkostrzewa commented:
I've added missing tests and rebased on top of the `master` branch.

If there will be any changes requested I can address them next wednesday at earliest.

anoopgarlapati commented on core/src/main/java/org/springframework/security/oauth2/server/authorization/web/DefaultOAuth2ClientAuthenticationConverter.java:
```diff
@@ -0,0 +1,72 @@
+		if (credentials.length != 2) {
+			throw new OAuth2AuthenticationException(new OAuth2Error(OAuth2ErrorCodes.INVALID_TOKEN));
+		}
+		return new OAuth2ClientAuthenticationToken(credentials[0], credentials[1]);
```
@pkostrzewa cc: @jgrandja 
As per [RFC 6749 - section 2.3.1](https://tools.ietf.org/html/rfc6749#section-2.3.1), the client identifier and client password are further encoded using `application/x-www-form-urlencoded` encoding algorithm and then used as username and password .
> The client identifier is encoded using the
   "application/x-www-form-urlencoded" encoding algorithm per
   Appendix B, and the encoded value is used as the username; the client
   password is encoded using the same algorithm and used as the
   password. 

There is an issue already logged in legacy oauth2 project spring-projects/spring-security-oauth#1826 and we should implement this according to specification in this project.

dfcoffin replied:
@anoopgarlapati
Although [RFC 6749 - section 2.3.1,](https://tools.ietf.org/html/rfc6749#section-2.3.1) indeed does say:

> The client identifier is encoded using the
> "application/x-www-form-urlencoded" encoding algorithm per
> Appendix B, and the encoded value is used as the username; the client
> password is encoded using the same algorithm and used as the
> password.

It further states:

> Including the client credentials in the request-body using the two parameters is NOT RECOMMENDED and SHOULD be limited to clients unable to directly utilize the HTTP Basic authentication scheme (or other password-based HTTP authentication schemes).  the parameters can only be transmitted in the request-body and MUST NOT be included in the request URI.
> 
> The authorization server MUST require the of TLS as described in [Section 1.6](https://tools.ietf.org/html/rfc6749#section-1.6) when sending requests using password authentication.
> 
> Since this client authentication method involves a password, the authorization server MUST protect any endpoint utilizing it against brute force attacks.

OrangeDog replied:
Those sections do not contradict each other. The encoding needs to be applied when using Basic authentication.

jgrandja commented:
Thank you for all the updates @pkostrzewa !

Overall, the implementation looked good, however, I did add a polish commit in order to get this merged.

Please take a look at the changes applied in the polish commit and let me know if you have any questions.

Thank you again and I look forward to another contribution 👍
